### PR TITLE
Allow setting if the critical hit should disable sweep attack in CriticalHitEvent, adding SweepAttackEvent

### LINKED ISSUE #1313: [Feature Request] allow `CriticalHitEvent` to not disable  sweep when set to criticalHit
Vanilla logic:
1. check if player can perform critical hit (flag1)
2. apply critical hit bonus
3. check if player can sweep attack (flag2), which requires player to not be able to critical hit in vanilla sense

Current NeoForge logic:
Directly modifies flag1 when mods enables critical hit. This disables sweep attack unintentionally. The suggested behavior is to use the vanilla critical hit flag for sweep attack check, which means changing the following code:

```
var critEvent = net.neoforged.neoforge.common.CommonHooks.fireCriticalHit(this, target, flag1, flag1 ? 1.5F : 1.0F);
flag1 = critEvent.isCriticalHit();
if (flag1) {
    f *= critEvent.getDamageMultiplier();
}

float f3 = f + f1;
boolean flag2 = false;
double d0 = (double)(this.walkDist - this.walkDistO);
if (flag4 && !flag1 && !flag && this.onGround() && d0 < (double)this.getSpeed()) {
    ItemStack itemstack1 = this.getItemInHand(InteractionHand.MAIN_HAND);
    flag2 = itemstack1.canPerformAction(net.neoforged.neoforge.common.ItemAbilities.SWORD_SWEEP);
}
```

to

```
var critEvent = net.neoforged.neoforge.common.CommonHooks.fireCriticalHit(this, target, flag1, flag1 ? 1.5F : 1.0F);
boolean disableSweep = flag1;
flag1 = critEvent.isCriticalHit();
if (flag1) {
    f *= critEvent.getDamageMultiplier();
}

float f3 = f + f1;
boolean flag2 = false;
double d0 = (double)(this.walkDist - this.walkDistO);
if (flag4 && !disableSweep && !flag && this.onGround() && d0 < (double)this.getSpeed()) {
    ItemStack itemstack1 = this.getItemInHand(InteractionHand.MAIN_HAND);
    flag2 = itemstack1.canPerformAction(net.neoforged.neoforge.common.ItemAbilities.SWORD_SWEEP);
}
```

Alternative solutions such as adding whether to disable sweep as part of `CriticalHitEvent` would also address this issue.

### PR DESCRIPTION
**Pull request by lcy0x1**
Rework of PR https://github.com/neoforged/NeoForge/pull/1314
Closes Issue https://github.com/neoforged/NeoForge/issues/1313

This PR does the following:
1. Adds `disableSweep` in `CriticalHitEvent` to track if sweep attack should be disabled. Legacy methods that enables critical hit will disable sweep attack. 
2. Adds a new method in `CriticalHitEvent` to enable critical hit without disabling sweep attack.
3. Adds a new method to set if the critical hit should disable sweep attack.
4. Adds `SweepAttackEvent` to determine if player should sweep attack, considering both vanilla conditions and `CriticalHitEvent.disableSweep()`

The vanilla logic for critical hit and sweep attack is:
- Check critical hit conditions, including:
  - attack is full strength
  - player is not sprinting
  - player is falling
  - player is not on ground
  - player is not on climbable blocks
  - player is not in water
  - player is not blinded
  - player is not a passenger
  - target is a living entity
- Check sweep attack conditions, including
  - attack is full strength
  - player is not sprinting
  - player is not reaching its full speed
  - player is on ground
  - this attack is not critical hit
- Lastly, check if the weapon support critical hit

This PR propose the following logic:
- `flag1` would represent all 9 vanilla critical hit conditions
- `allowSweep` would represent first 4 vanilla sweep conditions (excluding critical hit check)
- `CriticalHitEvent` determines if the attack would disable sweep
- `SweepAttackEvent` determines if the attack would be a sweep attack, considering `allowSweep` and `CriticalHit.disableSweep`

### CODE REVIEW

File: src/main/java/net/neoforged/neoforge/event/entity/player/CriticalHitEvent.java [hunk 1]
```diff
@@ -74,19 +73,36 @@ public boolean isCriticalHit() {
     }
 
     /**
-     * Changes the critical hit state.
-     * 
+     * Changes the critical hit state. If set critical hit to true, disables sweep attack
```
Comment (Shadows-of-Fire) [inline src/main/java/net/neoforged/neoforge/event/entity/player/CriticalHitEvent.java]: ~~If set critical hit to true, disables sweep attack~~ -> If set to true, `disableSweep` will also be set to true.

Comment (lcy0x1) [reply src/main/java/net/neoforged/neoforge/event/entity/player/CriticalHitEvent.java]: fixed

File: patches/net/minecraft/world/entity/player/Player.java.patch [hunk 2]
```diff
@@ -268,9 +268,16 @@
...
-@@ -1179,9 +_,7 @@
+                     boolean flag2 = false;
                      double d0 = (double)(this.walkDist - this.walkDistO);
-                     if (flag4 && !flag1 && !flag && this.onGround() && d0 < (double)this.getSpeed()) {
+-                    if (flag4 && !flag1 && !flag && this.onGround() && d0 < (double)this.getSpeed()) {
++                    boolean allowSweepPreconditions = flag4 && !flag && this.onGround() && d0 < (double)this.getSpeed();
++                    // Neo: the boolean local above (allowSweepPreconditions) is the vanilla sweep attack condition, excluding the condition of not doing critical hit
++                    // Fire the sweep attack event based on current information.
++                    // The sweep attack event default result will be allowSweep && !critEvent.disableSweep()
++                    // The event cannot bypass item ability check
```
Comment (Shadows-of-Fire) [inline patches/net/minecraft/world/entity/player/Player.java.patch]: Why is the event limited to not bypassing the item ability check? That seems like a perfectly reasonable thing for this to be able to override, especially given all the other things it's already able to bypass.

Comment (lcy0x1) [reply patches/net/minecraft/world/entity/player/Player.java.patch]: addressed

**Comment by KnightMiner** [comment]
I asked in the other PR, what is the behavior when that property is set on axes or other weapons with no sweep, I assume just ignored? Or am I able to force sweeps onto any weapon?

Personally I think it should just be ignored but wanted to double check so I know I dont have to deal with my weapons suddenly gaining sweep attacks.

**Comment by lcy0x1** [comment]
> I asked in the other PR, what is the behavior when that property is set on axes or other weapons with no sweep, I assume just ignored? Or am I able to force sweeps onto any weapon?
> 
> Personally I think it should just be ignored but wanted to double check so I know I dont have to deal with my weapons suddenly gaining sweep attacks.

Per Shadow’s request, it allows bypassing item ability check. Do you think I should have `enableSweep` and `enableSweepBypassItemAbilityCheck` separately?

**Review by Shadows-of-Fire — APPROVED** [review]

**Pull request merged**